luarules/gadgets/cmd_areaguard.lua

```diff
@@ -0,0 +1,430 @@
+end
+
+--------------------------------------------------------------------------------
+--------------------------------------------------------------------------------
+
+local spUtilities = Spring.Utilities
+--local SUC = spUtilities.CMD
+
+local env = getfenv()
+local SUC = VFS.Include("LuaRules/Configs/customcmds.lua", nil, VFS.GAME)
+for cmdName, cmdID in pairs(SUC) do
+	env["CMD_" .. cmdName] = cmdID
+end
+env.CMD_SETHAVEN = SUC.RETREAT_ZONE
+
+local CMD_AREA_GUARD = SUC.AREA_GUARD
+local CMD_ORBIT = SUC.ORBIT
+local CMD_ORBIT_DRAW = SUC.ORBIT_DRAW
+----------
+---------
+local GiveClampedMoveGoalToUnit = spUtilities.GiveClampedMoveGoalToUnit
+local Angle = VFS.Include("LuaRules/Utilities/vector.lua", nil, VFS.GAME)
+--local Angle = spUtilities.Vector.Angle
+
+local spGetUnitPosition = Spring.GetUnitPosition
+local spGetUnitVelocity = Spring.GetUnitVelocity
+local spAreTeamsAllied  = Spring.AreTeamsAllied
+local spGetUnitTeam     = Spring.GetUnitTeam
+--------------------------------------------------------------------------------
+--------------------------------------------------------------------------------
+-- Global Variables
+
+local areaGuardCmd = {
+	id      = CMD_AREA_GUARD,
+	name    = "Guard2",
+	action  = "areaguard",
+	cursor  = 'Guard',
+	type    = CMDTYPE.ICON_UNIT_OR_AREA,
+	tooltip = "Guard the unit or units",
+	--hidden  = true,
+}
+
+local orbitDrawCmd = {
+	id      = CMD_ORBIT_DRAW,
+	name    = "OrbitDraw",
+	action  = "orbitdraw",
+	cursor  = 'Guard',
+	type    = CMDTYPE.ICON_UNIT,
+	tooltip = "Circle around the unit in a protective manner",
+	hidden  = true,
+}
+
+-- ud.canGuard is true for units for which it should not be true. This table
+-- is generated as new unit types are created. It check for guard commands.
+local canGuardUnitDefIDs = {}
+
+local newGuards = {}
+local oldGuards = {}
+
+local oldcircuitTime = {}
+local newcircuitTime = {}
+local alreadyHandled = {}
+
+local keepOrder = {}
+local nextKeepOrder = {}
+local unitOrder = {}
+
+local circleDirection = {}
+local lastUpdateTime = {}
+
+local guardAngle = {}
+
+local frame = 0
+
+-- 2*pi comes from the circumference of the circle.
+-- 15 comes from the gap between move goals being recieved.
+local RAW_MOVE_MODE = false
+
+local CIRC_MULT = 15 * 2*math.pi
+
+local UPDATE_MULT = 2
+local PREDICTION = 30
+local UPDATE_GAP_LEEWAY = 30*UPDATE_MULT
+
+local RADIUS_BAND_SIZE = 100
+local ALLOW_CIRCLE_ENEMY = true
+
+local cos = math.cos
+local sin = math.sin
+local pi = math.pi
+
+local CMD_GUARD     = CMD.GUARD
+local CMD_STOP      = CMD.STOP
+local CMD_OPT_SHIFT = CMD.OPT_SHIFT
+local spGiveOrderToUnit = Spring.GiveOrderToUnit
+
+for i = 1, #UnitDefs do
+	local ud = UnitDefs[i]
+	canGuardUnitDefIDs[i] = not ud.isImmobile or ud.isFactory
+end
+
+--------------------------------------------------------------------------------
+--------------------------------------------------------------------------------
+-- Functions
+
+local function RandomPermutation(n)
```

Comment:
sounds like an excellent thing to put into table utilities, if not there already